tanjun/parsing.py

```diff
@@ -31,8 +31,321 @@
+        self.default = default
+        self.empty_value = empty_value
+        self._flags = dict(flags) if flags else {}
+        self._is_option = is_option
+        self.key = key
+        self.names = [name, *names]
+
+    @property
+    def converters(
+        self,
+    ) -> typing.AbstractSet[typing.Union[typing.Callable[[str], typing.Any], traits.Converter[typing.Any]]]:
+        return frozenset(self._converters)
+
+    @property
+    def flags(self) -> typing.MutableMapping[str, typing.Any]:
+        return self._flags
+
+    @property
+    def is_option(self) -> bool:
+        return self._is_option
+
+    def add_converter(
+        self, converter: typing.Union[typing.Callable[[str], typing.Any], traits.Converter[typing.Any]], /
+    ) -> None:
+        self._converters.add(converter)
+
+    def remove_converter(
+        self, converter: typing.Union[typing.Callable[[str], typing.Any], traits.Converter[typing.Any]], /
+    ) -> None:
+        self._converters.remove(converter)
+
+    def bind_component(self, component: traits.Component, /) -> None:
+        pass
+
+    async def convert(self, ctx: traits.Context, value: str) -> typing.Any:
+        sources: typing.MutableSequence[ValueError] = []
+        for converter in self._converters:
+            try:
+                if isinstance(converter, traits.Converter):
+                    return await converter.convert(ctx, value)
+
+                return converter(value)
+
+            except ValueError as exc:
+                sources.append(exc)
+
+        raise errors.ConversionError(sources)
+
+
+class ShlexParser(traits.Parser):
+    __slots__: typing.Sequence[str] = ("_arguments", "_options")
+
+    def __init__(self, *, parameters: typing.Optional[typing.Iterable[traits.Parameter]] = None) -> None:
+        self._arguments: typing.MutableSequence[traits.Parameter] = []
+        self._options: typing.MutableSequence[traits.Parameter] = []
+
+        if parameters is not None:
+            self.set_parameters(parameters)
+
+    @property
+    def parameters(self) -> typing.Sequence[traits.Parameter]:
+        return (*self._arguments, *self._options)
+
+    def add_parameter(self, parameter_: traits.Parameter, /) -> None:
+        if parameter_.is_option:
+            self._options.append(parameter_)
+
+        else:
+            self._arguments.append(parameter_)  # TODO: verify the signature
+
+    def remove_parameter(self, parameter_: traits.Parameter, /) -> None:
+        if parameter_.is_option:
+            self._options.remove(parameter_)
+
+        else:
+            self._arguments.remove(parameter_)  # TODO: verify the signature
+
+    def set_parameters(self, parameters: typing.Iterable[traits.Parameter], /) -> None:
+        self._arguments = [parameter_ for parameter_ in parameters if not parameter_.is_option]
+        self._options = [parameter_ for parameter_ in parameters if parameter_.is_option]
+        # TODO: verify the signature
```

Comment:
Might make verifying the signature a decorator method instead and add ”function” attribute to executable commands if it isn’t already there (will be optional to account for groups)

Reply:
Where all that’s verified here is basic details like how a greedy/multi argument must be the last argument